ArcGISRuntimeSDKQt_CppSamples/UtilityNetwork/ValidateUtilityNetworkTopology/ValidateUtilityNetworkTopology.cpp

```diff
@@ -0,0 +1,671 @@
+
+  //  Get the trace result from the utility network.
+  m_utilityNetwork->traceAsync(m_traceParameters).then(this, [this](QList<UtilityTraceResult*>)
+  {
+
+    m_busy = false;
+    emit isBusy();
+
+    UtilityTraceResult* result = m_utilityNetwork->traceResult()->at(0);
+
+    const QList<UtilityElement*> elements = static_cast<UtilityElementTraceResult*>(result)->elements(this);
+    m_message = QString("Trace completed: %1 elements found.").arg(elements.length());
+    emit messageChanged();
+
+    QueryParameters deviceParams;
+    QueryParameters lineParams;
+    QList<qint64> deviceObjIds;
+    QList<qint64> lineObjIds;
+
+    for (UtilityElement* item : elements)
+    {
+      if (item->networkSource()->name() == "Electric Distribution Device")
+        deviceObjIds.append(item->objectId());
+      else if (item->networkSource()->name() == "Electric Distribution Line")
+        lineObjIds.append(item->objectId());
+    }
+
+    deviceParams.setObjectIds(deviceObjIds);
+    lineParams.setObjectIds(lineObjIds);
+
+    m_deviceFeatureLayer->selectFeaturesAsync(deviceParams, SelectionMode::Add).then(this, [](FeatureQueryResult*)
+    {
+    });
+
+    m_lineFeatureLayer->selectFeaturesAsync(lineParams, SelectionMode::Add).then(this, [](FeatureQueryResult*)
+    {
+    });
+  }).onFailed(this, [this]()
+  {
+    m_message = QString("Trace failed.\n"
+        "Click 'Get State' to check the updated network state.");
+    emit messageChanged();
+
+    m_busy = false;
+    emit isBusy();
+  });
+
+  m_clearBtn = true;
+  emit clearBtn();
+}
+
+void ValidateUtilityNetworkTopology::getLabelDefinitions()
+{
+  // Visualize attribute editing using labels
+  m_deviceLabelDefinition = createDeviceLabelDefinition();
+  m_lineLabelDefinition = createLineLabelDefinition();
+  LayerListModel* layers = m_mapView->map()->operationalLayers();
+
+  for (Layer* layer : *layers)
+  {
+    if (layer->layerType() == LayerType::FeatureLayer)
+    {
+      m_featureLayer = qobject_cast<FeatureLayer*>(layer);
+
+      if(m_featureLayer->name() == "Electric Distribution Device")
+      {
+        m_deviceFeatureLayer = m_featureLayer;
+        m_deviceFeatureLayer->labelDefinitions()->append(m_deviceLabelDefinition);
+        m_deviceFeatureLayer->setLabelsEnabled(true);
+      }
+      else if(m_featureLayer->name() == "Electric Distribution Line")
+      {
+        m_lineFeatureLayer = m_featureLayer;
+        m_lineFeatureLayer->labelDefinitions()->append(m_lineLabelDefinition);
+        m_lineFeatureLayer->setLabelsEnabled(true);
+      }
+    }
+  }
+}
+
+void ValidateUtilityNetworkTopology::setupTraceParameters()
+{
+  // Trace with a subnetwork controller as default starting location
+  m_utilityAssetType = m_utilityNetwork
+                           ->definition()
+                           ->networkSource("Electric Distribution Device")
+                           ->assetGroup("Circuit Breaker")
+                           ->assetType("Three Phase");
```

Comment:
This is concise, but if anything goes wrong this will crash. Consider some minimal checking here.

Reply:
added